internal/dbtools/membership_enumeration.go

```diff
@@ -0,0 +1,311 @@
+	GROUP BY
+		group_memberships.user_id;`
+)
+
+// EnumeratedMembership represents a single user-to-group membership, which may be direct or indirect
+type EnumeratedMembership struct {
+	GroupID   string
+	Group     *models.Group
+	UserID    string
+	User      *models.User
+	IsAdmin   bool
+	ExpiresAt null.Time
+	Direct    bool
+}
+
+// GetMembershipsForUser returns a fully enumerated list of memberships for a user, optionally with sqlboiler's generated models populated
+func GetMembershipsForUser(ctx context.Context, db *sql.DB, userID string, shouldPopulateAllModels bool) ([]EnumeratedMembership, error) {
+	enumeratedMemberships := []EnumeratedMembership{}
+
+	err := queries.Raw(membershipsByUserQuery, userID).Bind(ctx, db, &enumeratedMemberships)
+	if err != nil {
+		if !errors.Is(err, sql.ErrNoRows) {
+			return []EnumeratedMembership{}, err
+		}
+	}
+
+	if shouldPopulateAllModels {
+		enumeratedMemberships, err = populateModels(ctx, db, enumeratedMemberships)
+		if err != nil {
+			return []EnumeratedMembership{}, err
+		}
+	}
+
+	return enumeratedMemberships, nil
+}
+
+// GetMembersOfGroup returns a fully enumerated list of memberships in a group, optionally with sqlboiler's generated models populated
+func GetMembersOfGroup(ctx context.Context, db *sql.DB, groupID string, shouldPopulateAllModels bool) ([]EnumeratedMembership, error) {
+	enumeratedMemberships := []EnumeratedMembership{}
+
+	err := queries.Raw(membershipsByGroupQuery, groupID).Bind(ctx, db, &enumeratedMemberships)
+	if err != nil {
+		if !errors.Is(err, sql.ErrNoRows) {
+			return []EnumeratedMembership{}, err
+		}
+	}
+
+	if shouldPopulateAllModels {
+		enumeratedMemberships, err = populateModels(ctx, db, enumeratedMemberships)
+		if err != nil {
+			return []EnumeratedMembership{}, err
+		}
+	}
+
+	return enumeratedMemberships, nil
+}
+
+// GetAllGroupMemberships returns a fully enumerated list of all memberships in the database, optionally with sqlboiler's generated models populated (use with caution, potentially lots of data)
+func GetAllGroupMemberships(ctx context.Context, db *sql.DB, shouldPopulateAllModels bool) ([]EnumeratedMembership, error) {
+	enumeratedMemberships := []EnumeratedMembership{}
+
+	err := queries.Raw(membershipsByGroupQuery).Bind(ctx, db, &enumeratedMemberships)
+	if err != nil {
+		if !errors.Is(err, sql.ErrNoRows) {
+			return []EnumeratedMembership{}, err
+		}
+	}
+
+	if shouldPopulateAllModels {
+		enumeratedMemberships, err = populateModels(ctx, db, enumeratedMemberships)
+		if err != nil {
+			return []EnumeratedMembership{}, err
+		}
+	}
+
+	return enumeratedMemberships, nil
+}
+
+func populateModels(ctx context.Context, db *sql.DB, memberships []EnumeratedMembership) ([]EnumeratedMembership, error) {
+	groupIDs := make([]string, 0)
+	userIDs := make([]string, 0)
+
+	for _, m := range memberships {
+		if !contains(groupIDs, m.GroupID) {
+			groupIDs = append(groupIDs, m.GroupID)
+		}
```

Comment:
did you think about using maps for these instead of looping over the whole list each time?